source/common/ur_pool_manager.hpp

```diff
@@ -0,0 +1,188 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ */
+
+#ifndef USM_POOL_MANAGER_HPP
+#define USM_POOL_MANAGER_HPP 1
+
+#include "ur_api.h"
+#include "ur_pool_manager.hpp"
+#include "ur_util.hpp"
+
+#include <functional>
+#include <vector>
+
+namespace usm {
+
+/// @brief describes an internal USM pool instance.
+struct pool_descriptor {
+    ur_usm_pool_handle_t poolHandle;
+
+    ur_context_handle_t hContext;
+    ur_device_handle_t hDevice;
+    ur_usm_type_t type;
+    bool deviceReadOnly;
+
+    static bool equal(const pool_descriptor &lhs, const pool_descriptor &rhs);
+    static std::size_t hash(const pool_descriptor &desc);
+    static std::pair<ur_result_t, std::vector<pool_descriptor>>
+    create(ur_usm_pool_handle_t poolHandle, ur_context_handle_t hContext);
+};
+
+static inline std::pair<ur_result_t, std::vector<ur_device_handle_t>>
+urGetSubDevices(ur_device_handle_t hDevice) {
+    size_t nComputeUnits;
+    auto ret = urDeviceGetInfo(hDevice, UR_DEVICE_INFO_MAX_COMPUTE_UNITS,
+                               sizeof(nComputeUnits), &nComputeUnits, nullptr);
+    if (ret != UR_RESULT_SUCCESS) {
+        return {ret, {}};
+    }
+
+    ur_device_partition_property_t properties[] = {
+        UR_DEVICE_PARTITION_EQUALLY, static_cast<intptr_t>(nComputeUnits), 0};
+
+    // Get the number of devices that will be created
+    uint32_t deviceCount;
+    ret = urDevicePartition(hDevice, properties, 0, nullptr, &deviceCount);
+    if (ret != UR_RESULT_SUCCESS) {
+        return {ret, {}};
+    }
+
+    std::vector<ur_device_handle_t> sub_devices(deviceCount);
+    ret = urDevicePartition(hDevice, properties,
+                            static_cast<uint32_t>(sub_devices.size()),
+                            sub_devices.data(), nullptr);
+    if (ret != UR_RESULT_SUCCESS) {
+        return {ret, {}};
+    }
+
+    return {UR_RESULT_SUCCESS, sub_devices};
+}
+
+inline std::pair<ur_result_t, std::vector<ur_device_handle_t>>
+urGetAllDevicesAndSubDevices(ur_context_handle_t hContext) {
+    size_t deviceCount;
+    auto ret = urContextGetInfo(hContext, UR_CONTEXT_INFO_NUM_DEVICES,
+                                sizeof(deviceCount), &deviceCount, nullptr);
+    if (ret != UR_RESULT_SUCCESS) {
+        return {ret, {}};
+    }
+
+    std::vector<ur_device_handle_t> devices(deviceCount);
+    ret = urContextGetInfo(hContext, UR_CONTEXT_INFO_DEVICES,
+                           sizeof(ur_device_handle_t) * deviceCount,
+                           devices.data(), nullptr);
+    if (ret != UR_RESULT_SUCCESS) {
+        return {ret, {}};
+    }
+
+    std::vector<ur_device_handle_t> devicesAndSubDevices;
+    std::function<void(ur_device_handle_t)> addPoolsForDevicesRec =
+        [&](ur_device_handle_t hDevice) {
+            devicesAndSubDevices.push_back(hDevice);
+            auto [ret, subDevices] = urGetSubDevices(hDevice);
+            if (ret != UR_RESULT_SUCCESS) {
+                throw ret;
```

Comment:
Wouldn't returning this value from `urGetAllDevicesAndSubDevices` be better?
Or should it not be done because `urGetSubDevices` might already have partitioned some devices?

Reply:
Your're right, I should return the status here. Done.